### PR TITLE
[Merged by Bors] - feat: `InnerProductSpace (SeparationQuotient E)`

### PR DESCRIPTION
**Pull request by eric-wieser**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/InnerProductSpace/Basic.lean [hunk 1]
```diff
@@ -2368,6 +2368,33 @@ theorem ContinuousLinearMap.reApplyInnerSelf_smul (T : E →L[𝕜] E) (x : E) {
 
 end ReApplyInnerSelf_Seminormed
 
+section SeparationQuotient
+variable [SeminormedAddCommGroup E] [InnerProductSpace 𝕜 E]
+
+theorem Inseparable.inner_eq_inner {x₁ x₂ y₁ y₂ : E}
+    (hx : Inseparable x₁ x₂) (hy : Inseparable y₁ y₂) :
+    inner x₁ y₁ = (inner x₂ y₂ : 𝕜) :=
+  ((hx.prod hy).map continuous_inner).eq
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/Basic.lean:2377]: Thanks @ADedecker for pointing out that this `map` and `.eq` API existed!

**Comment by ADedecker** [comment]
Sorry, could you explain quickly what is the relation between this and #16707 ? 

Also, @yoh-tanimoto, I see you have the GNS in mind, do you already have some working code ? I was planning on doing this before the summer, but I don't want to step on your toes.

**Comment by eric-wieser** [comment]
My thought here is that https://github.com/leanprover-community/mathlib4/pull/16707 could probably cover the equivalence between `SeparationQuotient` and the quotient module by the null-space submodule; but that we shouldn't take that as the definition, and should just directly lift the `inner` operation through the quotient as I did for `norm` in #17452, and as we already did for `dist`.

**Comment by ADedecker** [comment]
Ah, I see. My opinion on this is "just use `SeparationQuotient`", so I'm not even sure we really need to state the other construction. But we can discuss that somewhere else.

bors d+ 
(In case you want to wait on @yoh-tanimoto's review)

**Comment by eric-wieser** [comment]
bors merge

> so I'm not even sure we really need to state the other construction. But we can discuss that somewhere else.

I think it would be nice to state simply to show equivalence, but probably if @yoh-tanimoto has further goals it isn't actually necessary to reach them.

**Comment by yoh-tanimoto** [comment]
you two are very quick, and the code is so short, I'm impressed!

@ADedecker No, I don't have a single line of code of GNS. I'm not sure what I will work on next, so if you wish, please go ahead.

I don't know whether other constructions (by `lift` of `toDualMap`)? are needed, but implementing `lift`s as (continuous) linear maps is necessary, I think. For example, if you have an automorphism of a C^*-algebra preserving the state, in the GNS representation it is represented by a unitary operator. That should be the norm completion of the lift.

Should I use #16707 for this purpose, or open a new one?

**Pull request closed without merging**